### PR TITLE
Integration of the hidden Hopfield network to the storkey_hopfield_demo.py examples directory for the ilustrative comparison with MLP

### PR DESCRIPTION
**Pull request by aprotsenko24**
The refactored version of storkey_hopfield_demo.py demonstrates the advantage opf using the hopfield nodes in case of the data scarcity + high noise input. The architecture differs by one of the hidden layers, second one with 128 entries, to be fully-hopfield network. The chosen architecture can be varied by user, e.g. instead of having the linear hidden layers in-between, we can have fully-hopfield arm with 784 neurons and then linear layer that classifies these 784 updated latent states to get probabilities for 10 classes. The two seperate methods in the factory have been replaced with one make_model_factoryThe refactored version of storkey_hopfield_demo.py demonstrates the advantage opf using the hopfield nodes in case of the data scarcity + high noise input. The architecture differs by one of the hidden layers, second one with 128 entries, to be fully-hopfield network. The chosen architecture can be varied by user, e.g. instead of having the linear hidden layers in-between, we can have fully-hopfield arm with 784 neurons and then linear layer that classifies these 784 updated latent states to get probabilities for 10 classes by enabling --all-hopfield argument. This design choice (Fully-Hopfield arm) does not perform well as we scale up the training MNIST dataset or add the noise, it creates a large gap between the MLP and Hopfield accuracy due to incorrect architecture: the Hopfield network is suitable for retrieving the patterns given the partial/corrupted input, while MLP learn the features of the images. Thus, we train the Hopfield network on something that is not suitable for, which is feature extraction and so it disables Hopfield network to perform in what it is good at -- pattern recognition. Here are the runs for the enabled --all-hopfield argument: 

  RESULTS SUMMARY
    Hopfield strength: 2.0
    Trials: 5, Epochs: 20

       K    Noise    Hopfield%         MLP%     Delta%    p-value   Sig        d
      50      0.0 77.18+/-0.28 77.95+/-0.27      -0.77     0.0113     *   -1.985
      50      1.0 71.16+/-0.30 73.20+/-0.26      -2.04     0.0056     *   -2.430
      50      2.0 58.05+/-0.57 62.96+/-0.40      -4.91     0.0001     *   -7.916
     100      0.0 78.51+/-0.43 80.39+/-0.38      -1.88     0.0035     *   -2.750
     100      1.0 70.78+/-0.50 74.15+/-0.54      -3.37     0.0002     *   -5.705
     100      2.0 55.49+/-0.78 61.09+/-0.68      -5.59     0.0002     *   -5.873
     500      0.0 80.08+/-0.89 84.13+/-0.24      -4.04     0.0064     *   -2.338
     500      1.0 71.47+/-0.44 76.06+/-0.27      -4.59     0.0008     *   -4.127
     500      2.0 53.67+/-0.39 59.22+/-0.30      -5.55     0.0010     *   -3.824

  Delta Accuracy Heatmap (Hopfield - MLP, percentage points):

       K  n=0.0  n=1.0  n=2.0
      50  -0.8*  -2.0*  -4.9*
     100  -1.9*  -3.4*  -5.6*
     500  -4.0*  -4.6*  -5.6*

    * = significant at p<0.05 
    
    While using the Hopfield network as one of the hidden states, 128 by 128 weight matrix, we find that the hidden Hopfield layer performs better as the noise in the data increases and the K (number of the samples) increases (use the code documentation as the reference).

### CODE REVIEW

**Review by matthewbehrend — CHANGES_REQUESTED** [review]
The goal is to demonstrate Hopfield chaining and to test the hypothesis that memory is more powerful when chained together with depth. Such a demo highlights that Hopfield is a composable element of a network in FabricPC.

The PR designed:
`784 -> 784(StorkeyHopfield) -> 10(Linear)`
Collapsed network to a single hidden layer. Agents will frequently misinterpret design goals with Hopfield because the literature is full of examples where the Hopfield network is monolithic, i.e. a single DxD weight matrix. The accuracy drop is a natural outcome because the network lost depth. It's not deep learning.

Suggested upgrades: extend the depth and test the hypothesis that a higher fraction of nodes converted from linear to Hopfield produce cumulative accuracy gains. Create CLI arguments to compare each network or both versus the MLP.
```
Network:     Structure:
"MLP"        784 -> 64(Linear) -> 64(Linear)          -> 64(Linear) -> 10(Linear)
"1hopfield"  784 -> 64(Linear) -> 64(StorkeyHopfield) -> 64(Linear) -> 10(Linear)
"2hopfield"  784 -> 64(Linear) -> 64(StorkeyHopfield) -> 64(StorkeyHopfield) -> 10(Linear)
```

**Comment by aprotsenko24** [comment]
Hey Matt,

Thank you for your feedback. Actually, it was my idea of creating a single 784 by 784 matrix to see how the network would perform. Later, I realized it was mistake, as the accuracy was dropping and it was inappropriate architecture due to "learning" constraints of the hopfield network (it learn the patterns in the dataset, playing a role of "memory", not for the features recognition from the input -- it quickly fails at this as the dataset grows). The file compares the architecture similiar to your suggestion, but shallow one:

`Architecture for Hopfield NN: `<br>`
Input (784) -> Linear (128) -> Hopfield (128) -> Linear, Output (10)  `<br>`
Input (784) -> Linear (128) -> Linear (128) -> Linear, Output (10)
`

**Comment by matthewbehrend** [comment]
Thank you for your further development on this!
**Verdict: request changes.** The requested structure is delivered — the three
depth-4 / width-64 arms match the spec exactly, `--networks` selects variants
against an always-included MLP baseline, and the table-driven `ARCH_CONFIGS`
factory is clean and extensible. This demo meaningfully demonstrates experiment
the experiment features of the library as well. One correctness regression and one
architecture issue block merge. Changes revolve around applying the correct test
for the hypothesis, clarifying the hypothesis statement, and making a reusable
N-contrast experiment runner available in the library.
Further details in 
[storkey_hopfield_demo_pr_review.md](https://github.com/user-attachments/files/28868358/storkey_hopfield_demo_pr_review.md)

### Required

1. **Paired design is broken.** `run_multi_arm_trials` builds loaders once per
   trial and shares them across arms (demo:362), but `FewShotLoader.__iter__`
   increments `self._epoch`, which changes the shuffle seed each pass
   (dataloader.py:516-524). With E epochs, MLP trains on epoch-shuffles
   0..E-1, the next arm on E..2E-1, the third on 2E..3E-1: arms see different
   batch streams, and an arm's accuracy changes with the `--networks`
   selection. `ABExperiment` recreates loaders per arm to prevent exactly this
   (ab_experiment.py:369-377). Fix: add `reset()` to the loaders and call it
   per arm (also covers `FashionMnistLoader`/`MnistLoader` with
   `shuffle=True`, and lets `ABExperiment` drop its per-arm TFDS reload).
2. **Promote the N-arm runner into the library.** `run_multi_arm_trials`
   (demo:333-403) duplicates `ABExperiment`'s trial loop in an example file,
   including the `KeyError` text verbatim (demo:390-395 vs
   ab_experiment.py:323-328). Add `PlannedMultiContrastExperiment` to
   `fabricpc/experiments` (arms list; constructor-declared
   `contrasts=[(arm_name, arm_name), ...]` as the hypothesis family; per-arm
   loader `reset()`; a descriptive `delta(arm_a, arm_b)` helper for
   reported-only deltas); reimplement `ABExperiment` as a 2-arm wrapper so
   the four existing example callers are unchanged.
3. **Test the two planned contrasts, not every arm vs MLP.** The hypothesis
   is two ordered increments: substitution 1 (1hopfield−MLP) and
   substitution 2 (2hopfield−1hopfield). The demo instead tests each arm
   against MLP (demo:503-508): 2hopfield−MLP is the total effect, and the
   second increment — the claim "2hopfield consistently beats 1hopfield"
   (demo:68) — is never tested. Declare exactly these two contrasts and star
   only them (two-sided paired t-tests on the per-trial difference vectors —
   two-sided deliberately, so significant negative cells stay visible;
   existing `paired_ttest`/`cohens_d` apply unchanged); report 2hopfield−MLP
   descriptively (mean ± SE, no star); make the second heatmap
   2hopfield−1hopfield. No omnibus ANOVA needed for two pre-specified
   contrasts.
4. **Disambiguate the hypothesis in the docstring.** "Cumulative accuracy
   gains under data scarcity and input noise" (demo:5-8) reads as gains
   accumulating with scarcity/noise; the results paragraph's
   "data-abundance" (demo:65-67) then contradicts it. State it as: gains
   cumulative in the number of Linear→StorkeyHopfield substitutions, with the
   K×noise grid as a sensitivity analysis of the per-substitution gain (grows
   with noise, requires sufficient K, negative at K=10 / low noise).
5. **Fix the setup note.** tensorflow_datasets *is* declared — the `tfds`
   extra (pyproject.toml:47-50). Replace the note with
   `pip install -e ".[tfds]"`. `tensorflow-cpu` is unnecessary: both loaders
   already call `tf.config.set_visible_devices([], "GPU")`.
6. **Add `tests/test_experiments.py`** (package currently untested): loader
   `reset()` replay, arm-order independence of per-arm results, only declared
   contrasts tested, missing-metric `KeyError`, `ABExperiment` delegation.
   Use a synthetic loader, not TFDS.

### Minor

- Mark the per-cell stars as exploratory in the legend (16 cells × 2 planned
  contrasts = 32 uncorrected tests). No within-cell correction: the
  cumulative claim is a conjunction, so requiring both contrasts to reject
  at α=0.05 with positive deltas is already a level-α test of it.
- K=10 trains on 64 of 100 samples per epoch (drop-remainder,
  dataloader.py:526) — caveat the K=10 row or shrink the batch at low K.
- `make_model_factory`: raise `ValueError` if `layer_types[0] != "Linear"`
  instead of failing later with a shape error.
- Print the `* = significant` legend whenever stars print, not only with the
  2-D heatmap (demo:569).
- Default `--num_epochs 1` (demo:156) will not resemble the documented
  20-epoch results; raise it or state expected default output.
- Squash-merge: commit 9c70771's message describes an `--all-hopfield` flag
  and 128-wide design that no longer exist.

### Acceptance checks

- 2hopfield per-trial accuracies bit-identical between
  `--networks 2hopfield` and `--networks 1hopfield 2hopfield` at fixed seed.
- Results block regenerated with the corrected paired runner.
- `pytest tests/` green, including the new experiments tests.

**Comment by matthewbehrend** [comment]
Addition of reset machinery to all dataloaders entangles the experiment class design with data loaders and creates increased burden on defining new dataloaders. A better approach would be to address the root cause at the single data loader with a pain point of constructing multiple instances in memory - the fewshot dataloader design. Attached remediation plan.
[storkey_hopfield_pr_review.md](https://github.com/user-attachments/files/30110012/storkey_hopfield_pr_review.md)

**Comment by matthewbehrend** [comment]
Thank you for the changes made to address all of the review issues. With the dataloader refactored this is ready to merge.
This PR demonstrates StorkeyHopfield as a composable graph element: substituting Linear hidden layers produces accuracy gains that accumulate with the substitution count under input noise. Promotes the multi-arm experiment machinery into the library.

**Review by matthewbehrend — APPROVED** [review]
All issues addressed. Excellent demo.
